core/include/kernel/refcount.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2017, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#ifndef __KERNEL_REFCOUNT_H
+#define __KERNEL_REFCOUNT_H
+
+#include <atomic.h>
+
+/*
+ * Reference conter
+ *
+ * When val becomes 0 refcount_inc() refuses to change the value any longer
+ * and returns false.
```

Comment:
I find the wording a bit weird, a more neutral tone may be better: "When val is 0, `refcount_inc()` does not change the value and returns false.". You may add "Otherwise, it increments the value and returns true".